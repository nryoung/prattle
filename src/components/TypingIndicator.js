import React, { Component } from 'react';
import styled from 'styled-components';
import * as colors from '../styles/colors';
import withChatkit from '../containers/WithChatkit';

const StyledTypingIndicator = styled.div`
  position: fixed;
  bottom: 56px;
  width: 100vw;
  max-width: 100%;
  text-align: center;
`;

const TypingIndicatorMessage = styled.div`
  display: inline-block;
  border-radius: 8px;
  background-color: ${colors.miami};
  padding: 8px;
  transition: opacity 500ms linear;
  opacity: ${(props) => (props.isTyping ? 1 : 0)};
`;

class TypingIndicator extends Component {
  state = {
    isSomeoneTyping: false,
    userNameTyping: null,
  };

  componentDidMount() {
    const { chatkit } = this.props;
    chatkit.connect().then((currentUser) => {
      currentUser.subscribeToRoom({
        roomId: process.env.REACT_APP_ROOM,
        hooks: {
          onUserStartedTyping: (user) => {
            this.setState({ isSomeoneTyping: true, userNameTyping: user.name });
          },
          onUserStoppedTyping: (user) => {
            this.setState({ isSomeoneTyping: false });
          },
        },
      });
    });
  }

  render() {
    const { userNameTyping, isSomeoneTyping } = this.state;
    return (
      <StyledTypingIndicator>
        <TypingIndicatorMessage isTyping={isSomeoneTyping}>
          {userNameTyping} is typing!
        </TypingIndicatorMessage>
      </StyledTypingIndicator>
    );
  }
}

export { TypingIndicator };

export default withChatkit(TypingIndicator);
