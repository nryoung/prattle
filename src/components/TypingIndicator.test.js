import React from 'react';
import renderer from 'react-test-renderer';
import flushPromises from '../utils/FlushPromises';
import { TypingIndicator } from './TypingIndicator';

const props = {
  chatkit: {
    connect: jest.fn(
      () =>
        new Promise((resolve) =>
          resolve({
            subscribeToRoom: jest.fn(
              () =>
                new Promise((resolve) =>
                  resolve({ id: ':messageId', text: ':messageText' })
                )
            ),
          })
        )
    ),
  },
};

it('it renders', async () => {
  const component = renderer.create(<TypingIndicator {...props} />);
  await flushPromises();
  expect(component.toJSON()).toMatchSnapshot();
});
