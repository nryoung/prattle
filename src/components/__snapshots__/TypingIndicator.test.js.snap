// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`it renders 1`] = `
<div
  className="sc-bdVaJa hciYRu"
>
  <div
    className="sc-bwzfXH dXqCMg"
  >
     is typing!
  </div>
</div>
`;
